### PR TITLE
Primary key set NOT NULL by default.

### LINKED ISSUE #2778: Primary key set NOT NULL by default
## Enhancement
The primary key is NOT NULL by default, no need to explicitly specify.

```
create table orders (
    dt date ,
    order_id bigint NOT NULL,
    user_id int NOT NULL,
    merchant_id int NOT NULL,
) PRIMARY KEY (dt, order_id)
DISTRIBUTED BY HASH(order_id) BUCKETS 4;
```

Error 1064: Syntax error in line 6: ) PRIMARY KEY (dt, order_id) ^ Encountered: ) Expected: IDENTIFIER

expected: 
create table success.

### LINKED ISSUE #2855: primary key column default not null
## Feature request

**Is your feature request related to a problem? Please describe.**
primary key column default not null

### PR DESCRIPTION
**Pull request by Linkerist**
For database null constraint, for all key models, there are three choices in starrocks for user now:
1. explicit specify, null
2. explicit specify, not null
3. implicit specified, that is write nothing, the default behavior is null

Product manager @Dshadowzh what to do #2778 to make primary key set NOT NULL by default in the 3 implicit specified scenario, this PR implement this idea.

Implementation scheme:

According to this idea, and the null constraint upon primary key itself, that is key column must be not null.
For primary key column null constraint, when null specify, should report an error, when not null specify, should working as normal, when implicit specified, default behavior should be working as not null specify.

A question:
For now, just two value passed in process instance, a two state boolean value named isAllowNull to hold null constraint value, which can not tell the difference between implicit specified and explicit specify null, so we can not make a decision, to working as normal, or report an error, that is the question.

The solution: 
For opt_is_allow_null without keyword scenario, change it to null value. Add a boolean value named isAllowNullImplicitly indicate that null constraint implicit specified.
For now, use the isAllowNull() method to test the value of isAllowNull, after the opt_is_allow_null changed, for primary key, should use setPrimaryKeyNonNullable() to update the real value to `isAllowNull` in create table stage, if the null constraint was implicit specified, should set to false, which is not null.

before this PR:
```SQL
mysql> create table test (`pk` int, `col1` int) ENGINE=OLAP primary key (`pk`) COMMENT "OLAP" DISTRIBUTED BY HASH(`pk`) BUCKETS 1 PROPERTIES("replication_num" = "1", "storage_format" = "v2");
ERROR 1064 (HY000): primary key column[pk] cannot be nullable
mysql> create table test (`pk` int, `col1` int null) ENGINE=OLAP primary key (`pk`) COMMENT "OLAP" DISTRIBUTED BY HASH(`pk`) BUCKETS 1 PROPERTIES("replication_num" = "1", "storage_format" = "v2");
ERROR 1064 (HY000): primary key column[pk] cannot be nullable
mysql> create table test (`pk` int, `col1` int not null) ENGINE=OLAP primary key (`pk`) COMMENT "OLAP" DISTRIBUTED BY HASH(`pk`) BUCKETS 1 PROPERTIES("replication_num" = "1", "storage_format" = "v2");
ERROR 1064 (HY000): primary key column[pk] cannot be nullable
mysql> create table test (`pk` int not null, `col1` int) ENGINE=OLAP primary key (`pk`) COMMENT "OLAP" DISTRIBUTED BY HASH(`pk`) BUCKETS 1 PROPERTIES("replication_num" = "1", "storage_format" = "v2");
Query OK, 0 rows affected (0.15 sec)

mysql> desc test;
+-------+------+------+-------+---------+-------+
| Field | Type | Null | Key   | Default | Extra |
+-------+------+------+-------+---------+-------+
| pk    | INT  | No   | true  | NULL    |       |
| col1  | INT  | Yes  | false | NULL    |       |
+-------+------+------+-------+---------+-------+
2 rows in set (0.01 sec)

mysql> desc test all;
+-----------+---------------+-------+------+------+-------+---------+-------+
| IndexName | IndexKeysType | Field | Type | Null | Key   | Default | Extra |
+-----------+---------------+-------+------+------+-------+---------+-------+
| test      | PRIMARY_KEYS  | pk    | INT  | No   | true  | NULL    |       |
|           |               | col1  | INT  | Yes  | false | NULL    |       |
+-----------+---------------+-------+------+------+-------+---------+-------+
2 rows in set (0.00 sec)

mysql> show create table test;
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                            |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test  | CREATE TABLE `test` (
  `pk` int(11) NOT NULL COMMENT "",
  `col1` int(11) NULL COMMENT ""
) ENGINE=OLAP 
PRIMARY KEY(`pk`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`pk`) BUCKETS 1 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
); |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> show columns from test;
+-------+------+------+------+---------+---------+
| Field | Type | Null | Key  | Default | Extra   |
+-------+------+------+------+---------+---------+
| pk    | int  | NO   | YES  | NULL    |         |
| col1  | int  | YES  | NO   | NULL    | REPLACE |
+-------+------+------+------+---------+---------+
2 rows in set (0.01 sec)

mysql> create table test1 (`pk` int not null, `col1` int null) ENGINE=OLAP primary key (`pk`) COMMENT "OLAP" DISTRIBUTED BY HASH(`pk`) BUCKETS 1 PROPERTIES("replication_num" = "1", "storage_format" = "v2
");
Query OK, 0 rows affected (0.01 sec)

mysql> desc test1;
+-------+------+------+-------+---------+-------+
| Field | Type | Null | Key   | Default | Extra |
+-------+------+------+-------+---------+-------+
| pk    | INT  | No   | true  | NULL    |       |
| col1  | INT  | Yes  | false | NULL    |       |
+-------+------+------+-------+---------+-------+
2 rows in set (0.00 sec)

mysql> desc test1 all;
+-----------+---------------+-------+------+------+-------+---------+-------+
| IndexName | IndexKeysType | Field | Type | Null | Key   | Default | Extra |
+-----------+---------------+-------+------+------+-------+---------+-------+
| test1     | PRIMARY_KEYS  | pk    | INT  | No   | true  | NULL    |       |
|           |               | col1  | INT  | Yes  | false | NULL    |       |
+-----------+---------------+-------+------+------+-------+---------+-------+
2 rows in set (0.00 sec)

mysql> show create table test1;
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                             |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test1 | CREATE TABLE `test1` (
  `pk` int(11) NOT NULL COMMENT "",
  `col1` int(11) NULL COMMENT ""
) ENGINE=OLAP 
PRIMARY KEY(`pk`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`pk`) BUCKETS 1 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
); |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> show columns from test1;
+-------+------+------+------+---------+---------+
| Field | Type | Null | Key  | Default | Extra   |
+-------+------+------+------+---------+---------+
| pk    | int  | NO   | YES  | NULL    |         |
| col1  | int  | YES  | NO   | NULL    | REPLACE |
+-------+------+------+------+---------+---------+
2 rows in set (0.00 sec)

mysql> create table test2 (`pk` int not null, `col1` int not null) ENGINE=OLAP primary key (`pk`) COMMENT "OLAP" DISTRIBUTED BY HASH(`pk`) BUCKETS 1 PROPERTIES("replication_num" = "1", "storage_format" = "v2");
Query OK, 0 rows affected (0.01 sec)

mysql> desc test2;
+-------+------+------+-------+---------+-------+
| Field | Type | Null | Key   | Default | Extra |
+-------+------+------+-------+---------+-------+
| pk    | INT  | No   | true  | NULL    |       |
| col1  | INT  | No   | false | NULL    |       |
+-------+------+------+-------+---------+-------+
2 rows in set (0.00 sec)

mysql> desc test2 all;
+-----------+---------------+-------+------+------+-------+---------+-------+
| IndexName | IndexKeysType | Field | Type | Null | Key   | Default | Extra |
+-----------+---------------+-------+------+------+-------+---------+-------+
| test2     | PRIMARY_KEYS  | pk    | INT  | No   | true  | NULL    |       |
|           |               | col1  | INT  | No   | false | NULL    |       |
+-----------+---------------+-------+------+------+-------+---------+-------+
2 rows in set (0.00 sec)

mysql> show create table test2;
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                                 |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test2 | CREATE TABLE `test2` (
  `pk` int(11) NOT NULL COMMENT "",
  `col1` int(11) NOT NULL COMMENT ""
) ENGINE=OLAP 
PRIMARY KEY(`pk`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`pk`) BUCKETS 1 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
); |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)

mysql> show columns from test2;
+-------+------+------+------+---------+---------+
| Field | Type | Null | Key  | Default | Extra   |
+-------+------+------+------+---------+---------+
| pk    | int  | NO   | YES  | NULL    |         |
| col1  | int  | NO   | NO   | NULL    | REPLACE |
+-------+------+------+------+---------+---------+
2 rows in set (0.00 sec)

mysql>
```
after this PR:
```SQL
mysql> create table test (`pk` int, `col1` int) ENGINE=OLAP primary key (`pk`) COMMENT "OLAP" DISTRIBUTED BY HASH(`pk`) BUCKETS 1 PROPERTIES("replication_num" = "1", "storage_format" = "v2");
Query OK, 0 rows affected (0.10 sec)

mysql> desc test;
+-------+------+------+-------+---------+-------+
| Field | Type | Null | Key   | Default | Extra |
+-------+------+------+-------+---------+-------+
| pk    | INT  | No   | true  |         |       |
| col1  | INT  | Yes  | false | NULL    |       |
+-------+------+------+-------+---------+-------+
2 rows in set (0.00 sec)

mysql> desc test all;
+-----------+---------------+-------+------+------+-------+---------+-------+
| IndexName | IndexKeysType | Field | Type | Null | Key   | Default | Extra |
+-----------+---------------+-------+------+------+-------+---------+-------+
| test      | PRIMARY_KEYS  | pk    | INT  | No   | true  |         |       |
|           |               | col1  | INT  | Yes  | false | NULL    |       |
+-----------+---------------+-------+------+------+-------+---------+-------+
2 rows in set (0.00 sec)

mysql> show create table test;
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                            |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test  | CREATE TABLE `test` (
  `pk` int(11) NOT NULL COMMENT "",
  `col1` int(11) NULL COMMENT ""
) ENGINE=OLAP 
PRIMARY KEY(`pk`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`pk`) BUCKETS 1 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
); |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)

mysql> create table test (`pk` int null, `col1` int) ENGINE=OLAP primary key (`pk`) COMMENT "OLAP" DISTRIBUTED BY HASH(`pk`) BUCKETS 1 PROPERTIES("replication_num" = "1", "storage_format" = "v2");
ERROR 1064 (HY000): primary key column[pk] cannot be nullable
mysql> create table test (`pk` int null, `col1` int null) ENGINE=OLAP primary key (`pk`) COMMENT "OLAP" DISTRIBUTED BY HASH(`pk`) BUCKETS 1 PROPERTIES("replication_num" = "1", "storage_format" = "v2");
ERROR 1064 (HY000): primary key column[pk] cannot be nullable
mysql> create table test (`pk` int null, `col1` int not null) ENGINE=OLAP primary key (`pk`) COMMENT "OLAP" DISTRIBUTED BY HASH(`pk`) BUCKETS 1 PROPERTIES("replication_num" = "1", "storage_format" = "v2");
ERROR 1064 (HY000): primary key column[pk] cannot be nullable
mysql> create table test1 (`pk` int not null, `col1` int) ENGINE=OLAP primary key (`pk`) COMMENT "OLAP" DISTRIBUTED BY HASH(`pk`) BUCKETS 1 PROPERTIES("replication_num" = "1", "storage_format" = "v2");
Query OK, 0 rows affected (0.00 sec)

mysql> desc test1;
+-------+------+------+-------+---------+-------+
| Field | Type | Null | Key   | Default | Extra |
+-------+------+------+-------+---------+-------+
| pk    | INT  | No   | true  |         |       |
| col1  | INT  | Yes  | false | NULL    |       |
+-------+------+------+-------+---------+-------+
2 rows in set (0.00 sec)

mysql> desc test1 all;
+-----------+---------------+-------+------+------+-------+---------+-------+
| IndexName | IndexKeysType | Field | Type | Null | Key   | Default | Extra |
+-----------+---------------+-------+------+------+-------+---------+-------+
| test1     | PRIMARY_KEYS  | pk    | INT  | No   | true  |         |       |
|           |               | col1  | INT  | Yes  | false | NULL    |       |
+-----------+---------------+-------+------+------+-------+---------+-------+
2 rows in set (0.01 sec)

mysql> show create table test1;
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                             |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test1 | CREATE TABLE `test1` (
  `pk` int(11) NOT NULL COMMENT "",
  `col1` int(11) NULL COMMENT ""
) ENGINE=OLAP 
PRIMARY KEY(`pk`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`pk`) BUCKETS 1 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
); |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> show columns from test1;
+-------+------+------+------+---------+---------+
| Field | Type | Null | Key  | Default | Extra   |
+-------+------+------+------+---------+---------+
| pk    | int  | NO   | YES  |         |         |
| col1  | int  | YES  | NO   | NULL    | REPLACE |
+-------+------+------+------+---------+---------+
2 rows in set (0.00 sec)

mysql> create table test2 (`pk` int not null, `col1` int null) ENGINE=OLAP primary key (`pk`) COMMENT "OLAP" DISTRIBUTED BY HASH(`pk`) BUCKETS 1 PROPERTIES("replication_num" = "1", "storage_format" = "v2");
Query OK, 0 rows affected (0.01 sec)

mysql> desc test2;
+-------+------+------+-------+---------+-------+
| Field | Type | Null | Key   | Default | Extra |
+-------+------+------+-------+---------+-------+
| pk    | INT  | No   | true  |         |       |
| col1  | INT  | Yes  | false | NULL    |       |
+-------+------+------+-------+---------+-------+
2 rows in set (0.01 sec)

mysql> desc test2 all;
+-----------+---------------+-------+------+------+-------+---------+-------+
| IndexName | IndexKeysType | Field | Type | Null | Key   | Default | Extra |
+-----------+---------------+-------+------+------+-------+---------+-------+
| test2     | PRIMARY_KEYS  | pk    | INT  | No   | true  |         |       |
|           |               | col1  | INT  | Yes  | false | NULL    |       |
+-----------+---------------+-------+------+------+-------+---------+-------+
2 rows in set (0.00 sec)

mysql> show create table test2;
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                             |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test2 | CREATE TABLE `test2` (
  `pk` int(11) NOT NULL COMMENT "",
  `col1` int(11) NULL COMMENT ""
) ENGINE=OLAP 
PRIMARY KEY(`pk`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`pk`) BUCKETS 1 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
); |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> show columns from test2;
+-------+------+------+------+---------+---------+
| Field | Type | Null | Key  | Default | Extra   |
+-------+------+------+------+---------+---------+
| pk    | int  | NO   | YES  |         |         |
| col1  | int  | YES  | NO   | NULL    | REPLACE |
+-------+------+------+------+---------+---------+
2 rows in set (0.00 sec)

mysql> create table test3 (`pk` int not null, `col1` int not null) ENGINE=OLAP primary key (`pk`) COMMENT "OLAP" DISTRIBUTED BY HASH(`pk`) BUCKETS 1 PROPERTIES("replication_num"                               Query OK, 0 rows affected (0.01 sec)

mysql> desc test3;
+-------+------+------+-------+---------+-------+
| Field | Type | Null | Key   | Default | Extra |
+-------+------+------+-------+---------+-------+
| pk    | INT  | No   | true  |         |       |
| col1  | INT  | No   | false |         |       |
+-------+------+------+-------+---------+-------+
2 rows in set (0.00 sec)

mysql> desc test3 all;
+-----------+---------------+-------+------+------+-------+---------+-------+
| IndexName | IndexKeysType | Field | Type | Null | Key   | Default | Extra |
+-----------+---------------+-------+------+------+-------+---------+-------+
| test3     | PRIMARY_KEYS  | pk    | INT  | No   | true  |         |       |
|           |               | col1  | INT  | No   | false |         |       |
+-----------+---------------+-------+------+------+-------+---------+-------+
2 rows in set (0.00 sec)

mysql> show create table test3;
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                                 |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test3 | CREATE TABLE `test3` (
  `pk` int(11) NOT NULL COMMENT "",
  `col1` int(11) NOT NULL COMMENT ""
) ENGINE=OLAP 
PRIMARY KEY(`pk`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`pk`) BUCKETS 1 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
); |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> show columns from test3;
+-------+------+------+------+---------+---------+
| Field | Type | Null | Key  | Default | Extra   |
+-------+------+------+------+---------+---------+
| pk    | int  | NO   | YES  |         |         |
| col1  | int  | NO   | NO   |         | REPLACE |
+-------+------+------+------+---------+---------+
2 rows in set (0.00 sec)

mysql>
```
resolve #2778

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java [hunk 1]
```diff
@@ -109,6 +111,10 @@ public boolean isAllowNull() {
         return isAllowNull;
     }
 
+    public void setIsAllowNullForPrimaryKeySql() {
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java:114]: setPrimaryKeyNonNullable()

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java [hunk 2]
```diff
@@ -86,6 +86,7 @@ public DefaultValueDef(boolean isSet, Expr expr) {
     private final TypeDef typeDef;
     private AggregateType aggregateType;
     private boolean isKey;
+    private Boolean isAllowNullTristate;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java:89]: This name is a little strange.  need to comment why we need `isAllowNull `  and `isAllowNullTristate` both

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java (hunk 1 above)
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java:114]: Need to comment why PrimaryKey is special.

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java (hunk 2 above)
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java:89]: I prefer a enum to describe this.

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java (hunk 1 above)
Comment (Linkerist) [reply fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java]: ok, changed.

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java (hunk 2 above)
Comment (Linkerist) [reply fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java]: yes, changed, added

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java (hunk 1 above)
Comment (Linkerist) [reply fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java]: added.

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java (hunk 2 above)
Comment (Linkerist) [reply fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java]: do you mean the variable isPrimaryKeyAllowNull should be a enum? 
or mean that the method isAllowNull() should return a enum, and do the judgement logic for primary key every single places?

**Review by decster — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java [hunk 3]
```diff
@@ -95,12 +99,17 @@ public ColumnDef(String name, TypeDef typeDef) {
...
         this.name = name;
         this.typeDef = typeDef;
         this.isKey = isKey;
         this.aggregateType = aggregateType;
-        this.isAllowNull = isAllowNull;
+        if (isAllowNull == null) {
+            this.isAllowNull = true;
+            this.isAllowNullImplicitly = true;
+        } else {
+            this.isAllowNull = isAllowNull;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java:111]: Better to explicitly set `isAllowNullImplicitly` to `false`

Comment (Linkerist) [reply fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java:111]: ok, changed.

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java [hunk 4]
```diff
@@ -86,7 +86,11 @@ public DefaultValueDef(boolean isSet, Expr expr) {
     private final TypeDef typeDef;
     private AggregateType aggregateType;
     private boolean isKey;
-    private boolean isAllowNull;
+    // For primary key column null constraint, when null specify, should report an error, when not null specify, should working as normal, when implicit specified, default behavior should be working as not null specify.
+    // For now, just two value assign to variable `isAllowNull`, a two state boolean value named isAllowNull to hold null constraint value, which can not tell the difference between implicit specified and explicit specify null, so we can not make a decision, to working as normal, or report an error, that is the question.
+    // For opt_is_allow_null without keyword scenario, change it to null value. Add a boolean value named isAllowNullImplicitly indicate that null constraint implicit specified.
+    private boolean isAllowNullImplicitly = false;
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java:92]: isAllowNullImplicit is ok

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java [hunk 5]
```diff
@@ -86,7 +86,11 @@ public DefaultValueDef(boolean isSet, Expr expr) {
     private final TypeDef typeDef;
     private AggregateType aggregateType;
     private boolean isKey;
-    private boolean isAllowNull;
+    // For primary key column null constraint, when null specify, should report an error, when not null specify, should working as normal, when implicit specified, default behavior should be working as not null specify.
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java:89]: Primary-key column should obey the not-null constraint. When creating a table, the not-null constraint will add to the primary-key column default. If the user specifies NULL explicitly, semantics analysis will report an error.

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java [hunk 6]
```diff
@@ -86,7 +86,11 @@ public DefaultValueDef(boolean isSet, Expr expr) {
...
+    // For now, just two value assign to variable `isAllowNull`, a two state boolean value named isAllowNull to hold null constraint value, which can not tell the difference between implicit specified and explicit specify null, so we can not make a decision, to working as normal, or report an error, that is the question.
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java:90]: Now, `isAllowNull` is used to indicate a null constraint hold or not. Primary-key and non-primary-key columns obey different constraints, so the `isAllowNull` can not be assigned a default value.
Add a new variable name `isAllowNullImplicit` to indicate the message. If`isAllowNullImplicit=true`, it indicates the null constraint is obeyed implicitly.

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java [hunk 7]
```diff
@@ -109,6 +119,18 @@ public boolean isAllowNull() {
         return isAllowNull;
     }
 
+    // For database null constraint, for all key models, there are three choices in starrocks for user now(until 2.0 deb29af76ee749ed8e99ec002cce4881b507c230):
+    // 1. explicit specify, null
+    // 2. explicit specify, not null
+    // 3. implicit specified, that is write nothing, the default behavior is null
+    // Product manager @Dshadowzh what to do #2778 to make primary key set NOT NULL by default in the 3 implicit specified scenario.
+    // so, according to this proposal, for primary key, should use setPrimaryKeyNonNullable() to update the real value to `isAllowNull` in create table stage, if the null constraint was implicit specified, should set to false, which is not null.
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java:127]: So add a new setPrimaryKeyNonNullable() to set `isAllowNull` to be true for primary key columns.

**Review by wyb — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java [hunk 8]
```diff
@@ -86,7 +86,12 @@ public DefaultValueDef(boolean isSet, Expr expr) {
     private final TypeDef typeDef;
     private AggregateType aggregateType;
     private boolean isKey;
-    private boolean isAllowNull;
+    // Primary-key column should obey the not-null constraint. When creating a table, the not-null constraint will add to the primary-key column default. If the user specifies NULL explicitly, semantics analysis will report an error.
+    // Now, isAllowNull is used to indicate a null constraint hold or not. Primary-key and non-primary-key columns obey different constraints, so the isAllowNull can not be assigned a default value.
+    // Add a new variable name isAllowNullImplicit to indicate the message. IfisAllowNullImplicit=true, it indicates the null constraint is obeyed implicitly.
+    // For opt_is_allow_null without keyword scenario, change it to null value. Add a boolean value named isAllowNullImplicit indicate that null constraint implicit specified.
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java]: ```// For opt_is_allow_null without keyword scenario, change it to null value. Add a boolean value named isAllowNullImplicit indicate that null constraint implicit specified.``` is redundant.

**Review by chaoyli — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Review by decster — APPROVED** [review]

**Pull request merged**